File: Java/Generate_Parentheses.java
public class Solution {
    
    private List<String> list;
    public List<String> generateParenthesis(int n) {
        list = new ArrayList<String>();
        backtrack(n, 0, 0, "");
        return list;
    }
    
    private void backtrack(int n, int left, int right, String ret) {
        if(n == left && n == right) {
            list.add(ret);
            return;
        }
        
        if(left < n)
            backtrack(n, left + 1, right, ret+"(");
        if(left > right)
            backtrack(n, left, right+1, ret+")");
    }
}

File: Swift/generate-parentheses.swift
/**
 * Problem Link: https://leetcode.com/problems/generate-parentheses/
 */
class Solution {
    var pSet: Set<String>!
    func generateParenthesis(_ n: Int) -> [String] {
        pSet = Set<String>()
        generate(n, n, "")
        return Array(pSet)
    }
    
    private func generate(_ left: Int, _ right: Int, _ s: String) {
        if left == 0 && right == 0 {
            pSet.insert(s)
            return
        }
        if left == right {
            generate(left-1, right, s + "(")
        } else {
            if left > 0 {
                generate(left-1, right, s + "(")
            }
            if right > 0 {
                generate(left, right-1, s + ")")
            }
        }
    }
}
/**
 * https://leetcode.com/problems/generate-parentheses/
 * 
 * 
 */ 
// Date: Wed Jun 16 09:55:42 PDT 2021
class Solution {
    func generateParenthesis(_ n: Int) -> [String] {
        var result = Set<String>()
        
        func dfs(_ left: Int, _ right: Int, _ combine: String) {
            if left == n, right == n {
                result.insert(combine)
                return
            }
            
            if left < n {
                dfs(left + 1, right, combine + "(")
            }
            
            if right < n, right < left {
                dfs(left, right + 1, combine + ")")
            }
        }
        
        dfs(0, 0, "")
        
        return Array(result)
    }
}